glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,212 @@
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        let request = PingRequest::new(uuid, url, body);
+        queue.push_back(request);
+    }
+
+    /// Clear pending pings queue.
+    pub fn clear_ping_queue(&self) {
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        queue.clear();
+    }
+
+    /// Get the next `PingRequest` in queue.
+    pub fn get_next_ping(&self) -> Option<PingRequest> {
+        let mut queue = self
+            .queue
+            .write()
+            .expect("Can't write to pending pings queue.");
+        queue.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::metrics::PingType;
+    use crate::tests::new_glean;
+    use std::thread;
+    use std::time::Duration;
+
+    #[test]
+    fn test_doesnt_error_when_there_are_no_pending_pings() {
+        let (glean, _) = new_glean(None);
+
+        // Try and get the next request
+        let request = glean.upload_manager.get_next_ping();
+
+        // Verify request was not returned
+        assert!(request.is_none());
+    }
+
+    #[test]
+    fn test_returns_ping_request_when_there_is_one() {
+        let (mut glean, _) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, true);
+        glean.register_ping_type(&ping_type);
+
+        // Submit the ping to populate the pending_pings directory
+        glean.submit_ping(&ping_type).unwrap();
+
+        // Try and get the next request
+        let request = glean.upload_manager.get_next_ping();
+
+        // Verify request was returned
+        assert!(request.is_some());
+    }
+
+    #[test]
+    fn test_pings_not_sent_pings_are_loaded_on_init() {
+        let mut tempdir = tempfile::tempdir().unwrap();
+        let n = 10;
+
+        // Start glean one and submit multiple pings
+        {
+            // We give tempdir to the `new_glean` function...
+            let (mut glean, dir) = new_glean(Some(tempdir));
+            // And then we get it back once that function returns.
+            tempdir = dir;
+
+            // Register a ping for testing
+            let ping_type = PingType::new("test", true, true);
+            glean.register_ping_type(&ping_type);
+
+            // Submit the ping multiple times
+            for _ in 0..n {
+                glean.submit_ping(&ping_type).unwrap();
+            }
+        }
+
+        // Start glean again and verify that the submitted pings get enqueued
+        {
+            let (glean, _) = new_glean(Some(tempdir));
+
+            // Getting this thread to sleep for just 100ms is enough for the
+            // spawned thread to lock the queue, which guarantees that,
+            // when we try `get_next_ping`, we will only get it after we are done enqueueing.
+            thread::sleep(Duration::from_millis(100));
```

Comment:
This is not ideal and fails intermittently on CI, as you may notice. I discussed it with @badboy and we figured that:
1. Testing multi-threaded stuff is hard;
2. What could be done is:
    2.1. Have a flag that will say whether processing of the pending_pings dir is done;
    2.2. When trying to `get_next_ping` either: a) tell the client to wait in case not done processing; b) return a `PingRequest` if there is one; c) return `None` if there is none.

I have started working on that, and it seems like it will be quite a big change. Just for the record :)